### PR TITLE
planner, expression: support enum index scan

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- No release note

### CODE REVIEW

File: util/ranger/points.go [hunk 1]
```diff
@@ -433,6 +437,51 @@ func handleBoundCol(ft *types.FieldType, val types.Datum, op string) (types.Datu
...
+	res := make([]*point, 0, len(ft.Elems)*2)
+	appendPointFunc := func(d types.Datum) {
+		res = append(res, &point{value: d, excl: false, start: true})
+		res = append(res, &point{value: d, excl: false, start: false})
+	}
+
+	for i := range ft.Elems {
+		enum, err := types.ParseEnumValue(ft.Elems, uint64(i+1))
+		if err != nil {
+			continue
```
Comment (lzmhhh123) [inline util/ranger/points.go:450]: How about add an error log here?

File: util/ranger/ranger.go [hunk 2]
```diff
@@ -101,6 +101,16 @@ func convertPoint(sc *stmtctx.StatementContext, point *point, tp *types.FieldTyp
 			// Ignore the types.ErrOverflow when we convert TypeNewDecimal values.
 			// A trimmed valid boundary point value would be returned then. Accordingly, the `excl` of the point
 			// would be adjusted. Impossible ranges would be skipped by the `validInterval` call later.
+		} else if tp.Tp == mysql.TypeEnum && terror.ErrorEqual(err, types.ErrTruncated) {
+			// Ignore the types.ErrorTruncated when we convert TypeEnum values.
+			// We should cover Enum upper overflow, and convert to the biggest value.
+			if point.value.GetInt64() > 0 {
+				upperEnum, err := types.ParseEnumValue(tp.Elems, uint64(len(tp.Elems)))
```
Comment (lzmhhh123) [inline util/ranger/ranger.go:108]: Why not directly `Enum{}`?

Comment (wshwsh12) [reply util/ranger/ranger.go:108]: If the enum struct is enum("a","b","c"), the condition `e  < 5 ` can convert to `e <= 3`, but not `e <= 0`.
So we need distinguish upper or lower overflow, like int type.

File: util/ranger/points.go (hunk 1 above)
Comment (wshwsh12) [reply util/ranger/points.go]: addressed

**Review by lzmhhh123 — APPROVED** [review]

File: util/ranger/points.go [hunk 3]
```diff
@@ -433,6 +439,51 @@ func handleBoundCol(ft *types.FieldType, val types.Datum, op string) (types.Datu
...
 
+func handleEnumFromBinOp(sc *stmtctx.StatementContext, ft *types.FieldType, val types.Datum, op string) []*point {
+	res := make([]*point, 0, len(ft.Elems)*2)
+	appendPointFunc := func(d types.Datum) {
+		res = append(res, &point{value: d, excl: false, start: true})
+		res = append(res, &point{value: d, excl: false, start: false})
+	}
+
+	for i := range ft.Elems {
+		enum, err := types.ParseEnumValue(ft.Elems, uint64(i+1))
```
Comment (winoros) [inline util/ranger/points.go:450]: Can we use a tmp enum to avoid small objects allocation here?

Comment (winoros) [reply util/ranger/points.go:450]: small objects allocation is a huge impact for ranger when the SQL is rather simple.

Comment (wshwsh12) [reply util/ranger/points.go]: ok

**Review by winoros — APPROVED** [review]

**Pull request merged**